File: 1122Programming/leetcode/876. Middle of the Linked List/ans.cpp
/**
 * https://leetcode.com/problems/middle-of-the-linked-list/
 * Problem ID: 876. Middle of the Linked List
 *
 * Author: Ateto
 *
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        ListNode* quick = head;
        ListNode* slow = head;
        while(slow->next && quick->next) {
            slow = slow->next;
            quick = quick->next;
            if(quick->next) {
                quick = quick->next;
            }
        }
        return slow;
    }
};

File: 1122Programming/leetcode/876. Middle of the Linked List/ans.rs
// https://leetcode.com/problems/middle-of-the-linked-list
// Problem ID: 876. Middle of the Linked List
//
// Author: revival0728
//
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn middle_node(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut ptr = head.clone();
        let mut len = 0_usize;
        while let Some(cur) = ptr {
            ptr = cur.next;
            len += 1;
        }
        let mut ans = head.clone();
        for i in 0..len/2 {
            ans = if let Some(cur) = ans { cur.next } else { panic!("") }
        }
        ans
    }
}
